### ISSUE TITLE
Add 'qual_name' property to Nodes

### ISSUE THREAD

**Issue by nicoddemus** [pytest-dev/pytest#7365]
As per https://github.com/pytest-dev/pytest/pull/7265#issuecomment-637590707.

**Comment by Zac-HD** [comment]
Should we follow [PEP 3155](https://www.python.org/dev/peps/pep-3155/) and name it `__qualname__`?

**Comment by RonnyPfannschmidt** [comment]
No, nodes are not classes

**Comment by symonk** [comment]
@nicoddemus gonna add this if thats ok? gonna roll with a `quar_name` read only property?

**Comment by nicoddemus** [comment]
Sure, sounds good, thanks!

**Comment by symonk** [comment]
@nicoddemus just wrapping my head around nodes and collection in general, could you spare any advice? take the following with a pinch of salt, here is some `pseudo` code that will need to change a lot, but am I barking up the wrong tree here?

-> if a node has no parent, return the node.name (possibly? strip .py etc?)
-> if the node has a parent, add the node name and for every parent(s) parent (appendleft) the name? then just replace appropriately, things such as `.py` extensions and the likes?

To create something like:

`'test_node_qual_name_property0.mypkg.test_mod.test_it'` when referring to a request.node.qual_name etc

**Comment by nicoddemus** [comment]
> -> if a node has no parent, return the node.name (possibly? strip out .py etc?)

Yeah just returning the `node.name` without stripping seems the correct approach.

No need to do any stripping, we just need to traverse the nodes and build the nodeid (untested):

```python
@property
def qual_name(self):
   return "::".join(x.name for x in self.listchain())
```

**Comment by symonk** [comment]
thanks @nicoddemus i will open a draft shortly, theres a few things i need to suss out / additional queries, especially listchain returning `()` when functions are inside classes.

**Comment by nicoddemus** [comment]
[PEP-3155](https://www.python.org/dev/peps/pep-3155/) explicitly defines `__qualname__` for classes and functions only, while [qualified name in the glossary](https://docs.python.org/3/glossary.html) is defined as:

> A dotted name showing the “path” from a module’s global scope to a class, function or method defined in that module, as defined in PEP 3155.

So it by definition does not include the full module dotted path as commented in https://github.com/pytest-dev/pytest/pull/7265#issuecomment-637590707.

What should we do? Re-use _qualified name_ in tests/classes to include also the module qualname might be confusing to users.

**Comment by symonk** [comment]
In regards to #8058 we need to discuss this more, as per @nicoddemus' advice.  I think its maybe ok for modules etc to just join on `.` ? as previous posts outlined using dunder qualname was not an option in reality here due to what Nodes are, any advice on how to proceed?
